### DISCUSSION TITLE
Is it possible to use tab normally when suggestions aren't shown and to accept suggestions when they are?

### DISCUSSION THREAD

**Discussion by Roundlay** [zbirenbaum/copilot.lua#153] [Q&A] [answered]
`copilot.vim` allows me to use `tab` for both accepting Copilot suggestions when they're visible, and for increasing the indentation level of the cursor as per normal when Copilot suggestions aren't visible. Is there a way to achieve the same functionality with `copilot.lua`?

**Comment by Roundlay** [comment]
Strangely enough, binding `accept` to `"<C-i>"` lets me use tab for both normal indentation and for accepting Copilot suggestions whenever they're visible, but this doesn't seem to work for other keybindings. Is this expected behaviour? Couldn't find anything about this in the help docs.

**Comment by MunifTanjim** [comment, marked as answer]
You can have super-tab like behavior, e.g.: https://github.com/MunifTanjim/dotfiles/blob/6b5199346f7e96065d5e517e61e2d8768e10770d/private_dot_config/nvim/lua/plugins/cmp.lua#L48-L63


For example:

```lua
vim.keymap.set('i', '<Tab>', function()
  if require("copilot.suggestion").is_visible() then
    require("copilot.suggestion").accept()
  else
    vim.api.nvim_feedkeys(vim.api.nvim_replace_termcodes("<Tab>", true, false, true), "n", false)
  end
end, { desc = "Super Tab" })
```

**Reply by Roundlay** [reply]
Thank you 🙏

**Reply by carmilso** [reply]
Hi! Thanks for the suggestion! Where would you write those lines? I'm not sure if they should be inside a mapping overwrite or inside the copilot.lua config

**Reply by MunifTanjim** [reply]
Disable the default `accept` keymap by doing:

```lua
require("copilot").setup({
  -- other options
  suggestion = {
    -- other options
    keymap = {
      -- other keymaps
      accept = false
    }
  }
})
```

Then add your custom accept keymap anywhere in your config. You can add it right below your `require("copilot").setup` call.

**Comment by mikkybang** [comment]
Hey @MunifTanjim  I did this, but it still doesn't work can you please help
```lua
return {
  {
    "zbirenbaum/copilot.lua",
    cmd = "Copilot",
    event = "InsertEnter",
    config = function()
      require("copilot").setup({ suggestion = { auto_trigger = true, keymap = { accept = false } } })
      vim.keymap.set("i", "<Tab>", function()
        if require("copilot.suggestion").is_visible() then
          require("copilot.suggestion").accept()
        else
          vim.api.nvim_feedkeys(vim.api.nvim_replace_termcodes("<Tab>", true, false, true), "n", false)
        end
      end, { desc = "Super Tab" })
    end,
  },
}

**Reply by MunifTanjim** [reply]
What's the output for `:imap <Tab>`?

**Reply by mikkybang** [reply]
Here:
`i  <Tab>       * <Lua 529: ~/.local/share/nvim/lazy/LazyVim/lua/lazyvim/plugins/coding.lua:23>`

**Reply by MunifTanjim** [reply]
https://github.com/LazyVim/LazyVim/blob/879e29504d43e9f178d967ecc34d482f902e5a91/lua/lazyvim/plugins/coding.lua#L21-L27

LazyVim is overriding your keymap.

**Comment by schroedermatthias** [comment]
$XDG_CONFIG_HOME/nvim/lua/plugins/copilot.lua:

```
return {
  {
    "zbirenbaum/copilot.lua",
    cmd = "Copilot",
    event = "InsertEnter",
    opts = {
      panel = { enabled = false },
      keymap = { accept = false },
      auto_refresh = true,
      filetypes = {
        yaml = true,
        markdown = true,
        suggestion = { auto_trigger = true },
      },
      vim.keymap.set("i", "<Tab>", function()
        if require("copilot.suggestion").is_visible() then
          require("copilot.suggestion").accept()
        else
          vim.api.nvim_feedkeys(vim.api.nvim_replace_termcodes("<Tab>", true, false, true), "n", false)
        end
      end, { desc = "Tab for copilot" }),
    },
  },
}
```

`:imap <Tab>` shows "Tab for copilot". But still, I cannot accept using Tab. Any ideas?

Your help is much appreciated.

Matthias

**Reply by reybits** [reply]
Move the `keymap` section inside the `suggestion` section:
```lua
suggestion = {
  auto_trigger = true,
  keymap = { accept = false }
}
```
